### PR TITLE
build: set Python and dependency versions with an optional Poetry env

### PR DESCRIPTION
**Pull request by tianjianjiang**
I will use code review comments to describe what I've done for the PR.

~~Three~~Two side notes:
1. Python 3.7.11 is for Colab;~~Although python>=3.8.10 also supports Apple Silicon and BigSur, 3.9 syntax is nicer for Black formatting;~~
~~2. GitHub Action doesn't have a roadmap for Apple M1 VM, therefore `test.yml` isn't really for the case;~~
1. Poetry is optional for managing venv and dependencies, but syncing with requirements(-dev).txt must be done manually for the time being.

### CODE REVIEW

File: .github/workflows/test.yml
Comment (tianjianjiang) [inline .github/workflows/test.yml:31]: `test_train.py` took about 5 minutes to complete so excluded for now.

File: pyproject.toml
Comment (tianjianjiang) [inline pyproject.toml:12]: Enforcing `3.9` if using Poetry.

File: setup.py
Comment (tianjianjiang) [inline setup.py:14]: Enforcing `>=3.9.1` if using setuptools.

File: pyproject.toml
Comment (tianjianjiang) [inline pyproject.toml:12]: Enforcing Apple Silicon compatible py39 if using Poetry.

**Comment by timoschick** [comment]
@SaulLu Since you've been setting up our current workflows and are more experienced with this than I am, could you perform the review for this pull request?

File: pyproject.toml
Comment (SaulLu) [inline pyproject.toml:9]: Just a nitpick, I have the impression that it is not the same thing as what is put in `setup.py`. Is it voluntary? :slightly_smiling_face:

**Review by SaulLu — APPROVED** [review]
Thank you very much for working on these additions. :+1: 

I had heard good things about poetry but never used it personally until now. If you don't mind, I have two small questions about poetry: 
- Can you confirm that the information is currently doubled between `poetry.lock` and (`requirements.txt`, `requirements-dev.txt` and èsetup.py`)? If this is the case, it's not a problem for me at all but I think we might benefit from stated it somewhere (in comments or on the readme).
- If I'm not mistaken, I understand that poetry fixes the virtual env packages (in the `poetry.lock` file). Knowing that torch requires special installation depending on the accelerator used, do you know how to handle this with poetry? 

Finally, I just want to say that if we run experiments on Jean Zay, we may have to change these requirements (python 3.9, and the versions of external libraries).

File: .github/workflows/test.yml
Comment (SaulLu) [reply .github/workflows/test.yml:31]: Thank you very much for this addition!

Comment (tianjianjiang) [reply .github/workflows/test.yml:31]: @SaulLu You're very welcome! I hope this will help a little without too much intrusion...

**Comment by tianjianjiang** [comment]
> Thank you very much for working on these additions. 👍

My pleasure 👍

----
> * Can you confirm that the information is currently doubled between `poetry.lock` and (`requirements.txt`, `requirements-dev.txt` and èsetup.py`)?

Partially yes. Although poetry.lock covers requirements.txt, requirements-dev.txt, and their dependencies, Poetry doesn't work closely with setup.py and setuptools. Some more info about it below but feel free to ignore it since it isn't really a big issue for experiments.

<details>
  <summary>[Poetry, pip-tools, and setuptools]</summary>
  I was a fan of pip-tools because not only it calculates deps fast and accurately, but also flexible for separating dev requirements. Everything is controllable except everything is manual. 😛
  Poetry is based on pip-tools, but the price of automation is that it isn't as flexible as pip-tools is. In addition, it doesn't want to process setup.py, so there is an edge case when certain deps is defined in setup.py, Poetry won't resolve it. (It happened to me once: python-poetry/poetry#1287)
  Therefore, I sometimes keep both setup.[py|cfg] and poetry.lock, and then I will also have to remember to sync requirements.txt (or constraints.txt) manually, and then hope one day somebody will improve the situation...
</details>

----
> If this is the case, it's not a problem for me at all but I think we might benefit from stated it somewhere (in comments or on the readme).

I will update README as soon as possible. In short, Poetry is likely something auxiliary.

----
> * If I'm not mistaken, I understand that poetry fixes the virtual env packages (in the `poetry.lock` file). Knowing that torch requires special installation depending on the accelerator used, do you know how to handle this with poetry?

Poetry reads pyproject.toml and then generates poetry.lock. As long as torch and accelerator versions can be defined by requirements.txt, same thing can happen to pyproject.toml, and then Poetry can handle them (except for the edge case aforementioned in [Poetry, pip-tools, and setuptools], python-poetry/poetry#1287).

----
> Finally, I just want to say that if we run experiments on Jean Zay, we may have to change these requirements (python 3.9, and the versions of external libraries).

I also expect that to happen in the near future. For this PR, I guess we can at least settle with Python version. If none of us need Apple M1 and Jean Zay requires older Python, I can downgrade it right away.

File: pyproject.toml
Comment (SaulLu) [reply pyproject.toml:9]: Indeed, I was talking about the `author` / `authors` section.

Thanks for the link, I now understand that the section keys differ between `poetry` and `Setuptools`

**Comment by SaulLu** [comment]
Thank you very much for all these explanations! :+1: 

> I also expect that to happen in the near future. For this PR, I guess we can at least settle with Python version. If none of us need Apple M1 and Jean Zay requires older Python, I can downgrade it right away.

At the moment, it's a bit early to tell about JZ. We can leave the python 3.9 version and come back down if necessary later?

**Comment by tianjianjiang** [comment]
@SaulLu

> At the moment, it's a bit early to tell about JZ. We can leave the python 3.9 version and come back down if necessary later?
> 
> EDIT: I have the impression that the [WG that runs experiments for the big language model uses python=3.8](https://github.com/bigscience-workshop/bigscience/tree/master/jz/envs#creating-production-conda-env), It might be easier to go with this version, what do you think?

I thought so (for py38) but then intriguingly modeling-promptsource did [python 3.8 -> >=3.7](https://github.com/bigscience-workshop/promptsource/commit/9b91d27d50d0f82bce6922156240a1e3a1510ba0) for a reason that I don't know (I haven't asked). I'm gonna ask around and see if we can have a more definite answer.

**Comment by tianjianjiang** [comment]
@timoschick @SaulLu In case package dependencies also concern you with data tooling, some additional info before your meeting:
- I guess many of us want to use Colab, and there are some minor dependency conflicts
    ```
    torchvision 0.10.0+cu102 has requirement torch==1.9.0, but you have torch 1.8.1.
    torchtext 0.10.0 has requirement torch==1.9.0, but you have torch 1.8.1.
    ```
- Besides `torch==1.9.0+cu102`, Colab uses Python 3.7.11.
- Since `accelerate==1.4.0`, it begins to support DeepSpeed that I believe some WG(s) may be working on it.

I can fix the above situations easily, as long as no serious conflicts with any of our local or Jean Zay envs.

**Comment by tianjianjiang** [comment]
Another new (actually old but I just forgot about it) info: https://github.com/bigscience-workshop/evaluation/pull/58#issuecomment-903781001

**Comment by tianjianjiang** [comment]
Colab (CPU) snippet and output:
```bash
!git clone -b build-use_py39_for_apple_silicon_with_optional_poetry https://github.com/tianjianjiang/bigscience-metadata.git
%cd bigscience-metadata
!pip install -q -e .
!TOKENIZERS_PARALLELISM=false python -m pytest .
```

```
============================= test session starts ==============================
platform linux -- Python 3.7.11, pytest-3.6.4, py-1.10.0, pluggy-0.7.1
rootdir: /content/bigscience-metadata, inifile:
plugins: typeguard-2.7.1, hydra-core-1.0.6
collected 14 items                                                             

tests/test_get_dataloaders.py .......                                    [ 50%]
tests/test_metadata_utils.py .....                                       [ 85%]
tests/test_train.py ..                                                   [100%]

========================= 14 passed in 327.71 seconds ==========================
```

**Comment by tianjianjiang** [comment]
@SaulLu 
I'm gonna merge this soon if no further concerns.

**Comment by SaulLu** [comment]
I apologise for answering so late, I didn't answer at the time because I don't really have a solution and unforthunatly I still don't. 

I understand all your points and indeed it seems to me a very good idea to be compatible with google collaboratory.

I can try to work on JZ with the dependencies that arrange on google collaboratory (for the moment I use `python 3.8` and `torch 1.8.1`). I will re-send a PR if there is any incompatibility for any reason :) 

I'm taking the opportunity to make 2 discuss the need to change the version of 2 other packages:
- for H, I think we would need the V version to be able to use `OmegaConf.to_object` which would allow transforming `args` which was a `omegaconf.dictconfig.DictConfig` into another object whose hash is deterministic. This is usefull for `datasets` library caching
- `wandb` I would like to use at least the V version in which they have released a feature that allows to synchronize offline runs that are still in progress

**Comment by tianjianjiang** [comment]
Hi @SaulLu ,
> I apologise for answering so late, I didn't answer at the time because I don't really have a solution (to combine compatibility with google collaboratory and use the same dependencies as WG engineering on JZ) and unforthunatly I still don't.

No problems at all. I was also unable to reply in almost 2 weeks.
I did actually have a plan B but hasn't played with it for quite a while, so not really confident if it will still work.
The plan B is that we may be able to use ngrok to connect to Colab. There
Honestly I don't really have a strong preference towards 3.7 (or 3.8). The only reason is that I'm lazy about trying out the plan B, or plan C, D, or E, etc., such as replacing the whole python environment of Colab, applying TPU Research Cloud (TRC) program, or convincing my boss to use paid resources.

> I can try to work on JZ with the dependencies that arrange on google collaboratory (for the moment I use `python 3.8` and `torch 1.8.1`). I will re-send a PR if there is any incompatibility for any reason :)

I think `torch` here is probably a more critical dependency than python itself. I believe that python can still be `>=3.7` for our package so both JZ and Colab will be fine, but `torch` may have to be specific. I guess `torch 1.8.1` will be just fine, even though `accelerate`'s examples are using `torch 1.9` now. Let me make a quick test on Colab and get back to you ASAP.

> * `hydra_core`, I think we would need the `1.1.0` ...
> * `wandb` I would like to use at least the `0.10.32` ...

I will also try them out on Colab together with `torch`, `torch-xla`, and `cloud-tpu-client`.

**Comment by tianjianjiang** [comment]
So far the behavior of Colab TPU seems similar to my previous setup, except for a harmless warning:
```
bsmetadata/train.py:107: UserWarning: 
config_path is not specified in @hydra.main().
See https://hydra.cc/docs/next/upgrades/1.0_to_1.1/changes_to_hydra_main_config_path for more information.
  @hydra.main(config_name="config")
```

**Comment by SaulLu** [comment]
Thank you very much for checking it out! :raised_hands: 

> The only reason is that I'm lazy about trying out the plan B, or plan C, D, or E, etc., such as replacing the whole python environment of Colab, applying TPU Research Cloud (TRC) program, or convincing my boss to use paid resources.
I understand perfectly, and I absolutely agree: the simpler the better! 

Concerning the error, I had to install an even newer version and had to correct this warning. I'm making a PR just after this one is merged to share this (very minor) change. :blush:

File: pyproject.toml
Comment (tianjianjiang) [inline pyproject.toml:18]: The semi-final changes for review.

File: requirements.txt
Comment (tianjianjiang) [inline requirements.txt:6]: Same as https://github.com/bigscience-workshop/metadata/pull/19/files#r702606374
See https://python-poetry.org/docs/dependency-specification/ for translating Poetry's `^` to pip requirements.txt.

**Review by timoschick — APPROVED** [review]

**Review by SaulLu — APPROVED** [review]
Thanks @tianjianjiang !

**Pull request merged**
